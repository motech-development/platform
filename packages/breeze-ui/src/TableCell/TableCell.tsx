/* eslint-disable react/jsx-props-no-spreading */
import React, { FC, memo, TdHTMLAttributes } from 'react';
import styled from 'styled-components';

const baseStyles = (noWrap: boolean, align = 'inherit') => `
  padding: 10px;
  text-align: ${align};
  vertical-align: middle;
  ${noWrap ? 'white-space: nowrap;' : ''}
`;

interface ITableCell {
  $noWrap: boolean;
}

const TableDataCell = styled.td<ITableCell>`
  ${({ align, $noWrap }) => `
    ${baseStyles($noWrap, align)}
    font-weight: 300;
  `}
`;

const TableHeadCell = styled.th<ITableCell>`
  ${({ align, $noWrap }) => `
    ${baseStyles($noWrap, align)}
    font-family: 'Cabin', sans-serif;
    font-weight: 600;
  `}
`;

export interface ITableCellProps
  extends TdHTMLAttributes<HTMLTableDataCellElement> {
  as?: 'td' | 'th';
  noWrap?: boolean;
}

const TableCell: FC<ITableCellProps> = ({
  as = 'td',
  noWrap = true,
  ...rest
}) =>
  as === 'td' ? (
    <TableDataCell $noWrap={noWrap} {...rest} />
  ) : (
    <TableHeadCell $noWrap={noWrap} {...rest} />
  );

export default memo(TableCell);
