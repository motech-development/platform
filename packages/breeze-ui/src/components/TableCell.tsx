import { FC, forwardRef, ThHTMLAttributes } from 'react';
import { Box } from 'react-polymorphic-box';
import TAlignment from '../utils/alignment';
import { classNames, textAlignmentClass } from '../utils/className';

type TTableCellAs = 'td' | 'th';

export interface ITableCellProps
  extends ThHTMLAttributes<HTMLTableDataCellElement> {
  align?: TAlignment;
  as?: TTableCellAs;
  noWrap?: boolean;
}

const TableCell: FC<ITableCellProps> = forwardRef<
  HTMLTableDataCellElement,
  ITableCellProps
>(
  (
    { align = 'left', as = 'td', className = '', noWrap = true, ...rest },
    ref,
  ) => (
    <Box
      as={as}
      className={classNames(
        'px-6 py-3',
        className,
        as === 'td' ? 'bg-white bg-opacity-70' : '',
        noWrap ? 'whitespace-nowrap' : '',
        textAlignmentClass(align),
      )}
      ref={ref}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
    />
  ),
);

export default TableCell;
