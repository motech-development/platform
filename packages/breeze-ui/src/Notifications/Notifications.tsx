import { faAsterisk, faBell } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { ReactNode, useEffect, useRef, useState } from 'react';
import { usePopper } from 'react-popper';
import styled from 'styled-components';
import Button from '../Button/Button';
import DataTable from '../DataTable/DataTable';
import useOutsideClick from '../hooks/useOutsideClick';
import TableCell from '../TableCell/TableCell';

const NotificationArrow = styled.div`
  height: 0;
  position: absolute;
  width: 0;
  border-bottom: 5px solid #007fa8;
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  top: -5px;
`;

const NotificationButton = styled(Button)`
  background: none;
  color: inherit;

  :hover {
    background: none;
    color: inherit;
  }
`;

const NotificationContainer = styled.div`
  box-shadow: 0 3px 5px -1px rgba(0, 0, 0, 0.2), 0 5px 8px 0 rgba(0, 0, 0, 0.14),
    0 1px 14px 0 rgba(0, 0, 0, 0.12);
`;

const NotificationWrapper = styled.div`
  position: relative;
`;

const AlertWrapper = styled.div`
  background-color: rgb(199, 56, 79);
  border-radius: 50%;
  color: #fff;
  padding: 5px;
  pointer-events: none;
  position: absolute;
  right: 5px;
  top: -5px;
  transform: scale(0.5);
  z-index: 1;
`;

const NotificationAlert = () => (
  <AlertWrapper>
    <FontAwesomeIcon icon={faAsterisk} />
  </AlertWrapper>
);

type Placement = 'bottom' | 'bottom-end' | 'bottom-start';

export interface INotificationsProps<T> {
  alert: boolean;
  cols?: number;
  items: T[];
  label: string;
  noResults: ReactNode;
  placement?: Placement;
  row: (item: T) => ReactNode;
  onClose(): void | Promise<void>;
}

function Notifications<T>({
  alert,
  cols = 1,
  items,
  label,
  noResults,
  onClose,
  placement = 'bottom',
  row,
}: INotificationsProps<T>) {
  const isFirstRun = useRef(true);
  const [visible, setVisible] = useState(false);
  const [referenceElement, setReferenceElement] =
    useState<HTMLDivElement | null>(null);
  const [arrowElement, setArrowElement] = useState<HTMLDivElement | null>(null);
  const [popperElement, setPopperElement] =
    useState<HTMLDivElement | null>(null);
  const { attributes, styles } = usePopper(referenceElement, popperElement, {
    modifiers: [
      {
        name: 'arrow',
        options: {
          element: arrowElement,
        },
      },
    ],
    placement,
    strategy: 'fixed',
  });
  const toggleNotifications = () => {
    setVisible(!visible);
  };

  useOutsideClick(referenceElement, () => {
    setVisible(false);
  });

  useEffect(() => {
    if (isFirstRun.current) {
      isFirstRun.current = false;

      return;
    }

    if (!visible) {
      (async () => {
        await Promise.resolve(onClose());
      })();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [visible]);

  return (
    <NotificationWrapper ref={setReferenceElement}>
      {alert && <NotificationAlert />}

      <NotificationButton
        aria-label={label}
        size="sm"
        onClick={toggleNotifications}
      >
        <FontAwesomeIcon icon={faBell} />
      </NotificationButton>

      {visible && (
        <NotificationContainer
          ref={setPopperElement}
          style={styles.popper}
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...attributes.popper}
        >
          <DataTable
            header={
              <TableCell as="th" align="left" colSpan={cols}>
                {label}
              </TableCell>
            }
            items={items}
            noResults={noResults}
            row={row}
          />

          <NotificationArrow ref={setArrowElement} style={styles.arrow} />
        </NotificationContainer>
      )}
    </NotificationWrapper>
  );
}

export default Notifications;
