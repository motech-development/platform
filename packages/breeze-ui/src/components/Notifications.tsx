import { Popover } from '@headlessui/react';
import { BellIcon } from '@heroicons/react/outline';
import { StarIcon } from '@heroicons/react/solid';
import { Placement } from '@popperjs/core';
import { ReactNode, useCallback, useEffect, useRef, useState } from 'react';
import { usePopper } from 'react-popper';
import { classNames, themeClass } from '../utils/className';
import TTheme from '../utils/theme';
import DataTable from './DataTable';
import TableCell from './TableCell';

export interface INotificationsProps<T> {
  alert: boolean;
  alertText?: string;
  colour?: TTheme;
  cols?: number;
  items: T[];
  label: string;
  noResults: ReactNode;
  placement?: Placement;
  row: (item: T) => ReactNode;
  onClose: () => void | Promise<void>;
}

function Notifications<T>({
  alert,
  alertText = 'New',
  colour = 'secondary',
  cols = 1,
  items,
  label,
  noResults,
  onClose,
  placement = 'bottom',
  row,
}: INotificationsProps<T>) {
  const isFirstRun = useRef(true);
  const [referenceElement, setReferenceElement] =
    useState<HTMLButtonElement | null>();
  const [popperElement, setPopperElement] = useState<HTMLDivElement | null>();
  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    modifiers: [
      {
        name: 'offset',
        options: {
          offset: [0, 18],
        },
      },
    ],
    placement,
  });
  const onCloseCallback = useCallback(() => {
    if (isFirstRun.current) {
      isFirstRun.current = false;

      return;
    }

    if (!popperElement) {
      (async () => {
        await Promise.resolve(onClose());
      })();
    }
  }, [onClose, popperElement]);

  useEffect(onCloseCallback, [onCloseCallback]);

  return (
    <Popover>
      {/* @tailwind: bg-blue-900 text-blue-400 focus:ring-offset-blue-900 hover:text-blue-100 focus:ring-blue-500 */}
      {/* @tailwind: bg-gray-900 text-gray-400 focus:ring-offset-gray-900 hover:text-gray-100 focus:ring-gray-500 */}
      {/* @tailwind: bg-green-900 text-green-400 focus:ring-offset-green-900 hover:text-green-100 focus:ring-green-500 */}
      {/* @tailwind: bg-red-900 text-red-400 focus:ring-offset-red-900 hover:text-red-100 focus:ring-red-500 */}
      {/* @tailwind: bg-yellow-900 text-yellow-400 focus:ring-offset-yellow-900 hover:text-yellow-100 focus:ring-yellow-500 */}
      <Popover.Button
        className={classNames(
          'p-1 rounded-full focus:outline-none focus:ring-2 focus:ring-offset-2 relative',
          themeClass(
            colour,
            'bg-{theme}-900 text-{theme}-400 focus:ring-offset-{theme}-900 hover:text-{theme}-100 focus:ring-{theme}-500',
          ),
        )}
        ref={setReferenceElement}
      >
        <span className="sr-only">{label}</span>

        <BellIcon className="h-6 w-6" aria-hidden="true" />

        {alert && (
          <span className="rounded-full text-sm font-medium bg-red-800 text-red-100 absolute top-0 right-0 p-0.5">
            <span className="sr-only">({alertText})</span>

            <StarIcon className="h-2.5 w-2.5" aria-hidden="true" />
          </span>
        )}
      </Popover.Button>

      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <Popover.Panel
        ref={setPopperElement}
        style={styles.popper}
        {...attributes.popper}
      >
        <DataTable
          header={
            <TableCell as="th" align="center" colSpan={cols}>
              {label}
            </TableCell>
          }
          items={items}
          noResults={noResults}
          row={row}
        />
      </Popover.Panel>
    </Popover>
  );
}

export default Notifications;
