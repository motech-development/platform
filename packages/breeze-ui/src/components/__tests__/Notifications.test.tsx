import { act, fireEvent, render, waitFor } from '@testing-library/react';
import { ReactNode } from 'react';
import TableCell from '../TableCell';
import Notifications from '../Notifications';

describe('Notifications', () => {
  let items: string[];
  let label: string;
  let noResults: ReactNode;
  let row: (item: string) => ReactNode;
  let onClose: jest.Mock;

  beforeEach(() => {
    items = ['Alert 1', 'Alert 2'];
    label = 'Notifications';
    noResults = <div>No results</div>;
    row = (item) => <TableCell>{item}</TableCell>;
    onClose = jest.fn();
  });

  it('should show an alert notification', () => {
    const { container } = render(
      <Notifications
        alert
        items={items}
        label={label}
        noResults={noResults}
        row={row}
        onClose={onClose}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not show an alert notification', () => {
    const { container } = render(
      <Notifications
        alert={false}
        items={items}
        label={label}
        noResults={noResults}
        row={row}
        onClose={onClose}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should display notifications when you click the alert button', async () => {
    const { container, findByRole } = render(
      <Notifications
        alert
        items={items}
        label={label}
        noResults={noResults}
        row={row}
        onClose={onClose}
      />,
    );

    await act(async () => {
      const button = await findByRole('button');

      fireEvent.click(button);
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should hide the notifications if you click elsewhere', async () => {
    const { container, findByRole, queryByRole } = render(
      <div>
        <Notifications
          alert
          items={items}
          label={label}
          noResults={noResults}
          placement="bottom-start"
          row={row}
          onClose={onClose}
        />
      </div>,
    );

    await act(async () => {
      const button = await findByRole('button');

      fireEvent.click(button);
    });

    await expect(findByRole('table')).resolves.toBeInTheDocument();

    act(() => {
      const body = container.firstChild as ChildNode;

      fireEvent.mouseDown(body);
    });

    expect(queryByRole('table')).not.toBeInTheDocument();
  });

  it('should display the label', async () => {
    const { findByRole, findAllByText } = render(
      <Notifications
        alert
        items={items}
        label={label}
        noResults={noResults}
        row={row}
        onClose={onClose}
      />,
    );
    const button = await findByRole('button');

    expect(button).toHaveTextContent('Notifications');

    fireEvent.click(button);

    await waitFor(() =>
      expect(findAllByText('Notifications')).resolves.toHaveLength(2),
    );
  });

  it('should call onClose when closed', async () => {
    const { findByRole } = render(
      <Notifications
        alert
        items={items}
        label={label}
        noResults={noResults}
        row={row}
        onClose={onClose}
      />,
    );

    await act(async () => {
      const button = await findByRole('button');

      fireEvent.click(button);

      fireEvent.click(button);
    });

    await waitFor(() => expect(onClose).toHaveBeenCalledTimes(1));
  });

  it('should call onClose when you click elsewhere', async () => {
    const { container, findByRole } = render(
      <div>
        <Notifications
          alert
          cols={1}
          items={items}
          label={label}
          noResults={noResults}
          row={row}
          onClose={onClose}
        />
      </div>,
    );

    await act(async () => {
      const button = await findByRole('button');
      const body = container.firstChild as ChildNode;

      fireEvent.click(button);

      fireEvent.mouseDown(body);
    });

    await waitFor(() => expect(onClose).toHaveBeenCalledTimes(1));
  });
});
