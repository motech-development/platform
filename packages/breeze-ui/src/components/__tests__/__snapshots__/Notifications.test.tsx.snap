// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications should display notifications when you click the alert button 1`] = `
<div>
  <button
    aria-controls="headlessui-popover-panel-6"
    aria-expanded="true"
    class="p-1 rounded-full focus:outline-none focus:ring-2 focus:ring-offset-2 relative bg-gray-900 text-gray-400 focus:ring-offset-gray-900 hover:text-gray-100 focus:ring-gray-500"
    id="headlessui-popover-button-5"
    type="button"
  >
    <span
      class="sr-only"
    >
      Notifications
    </span>
    <svg
      aria-hidden="true"
      class="h-6 w-6"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15 17h5l-1.405-1.405A2.032 2.032 0 0118 14.158V11a6.002 6.002 0 00-4-5.659V5a2 2 0 10-4 0v.341C7.67 6.165 6 8.388 6 11v3.159c0 .538-.214 1.055-.595 1.436L4 17h5m6 0v1a3 3 0 11-6 0v-1m6 0H9"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
    <span
      class="rounded-full text-sm font-medium bg-red-800 text-red-100 absolute top-0 right-0 p-0.5"
    >
      <span
        class="sr-only"
      >
        (
        New
        )
      </span>
      <svg
        aria-hidden="true"
        class="h-2.5 w-2.5"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
        />
      </svg>
    </span>
  </button>
  <div
    data-popper-escaped="true"
    data-popper-placement="bottom"
    data-popper-reference-hidden="true"
    id="headlessui-popover-panel-6"
    style="position: absolute; left: 0px; top: 0px; transform: translate(0px, 18px);"
  >
    <div
      class="flex flex-col"
    >
      <div
        class="-my-2 overflow-x-auto sm:-mx-6 lg:-mx-8"
      >
        <div
          class="py-2 align-middle inline-block min-w-full sm:px-6 lg:px-8"
        >
          <div
            class="shadow overflow-hidden border-b border-gray-200"
          >
            <table
              class="min-w-full table-auto"
            >
              <thead
                class="font-display font-semibold"
              >
                <tr
                  class="border-b-2 bg-blue-700 border-blue-800 text-blue-100"
                >
                  <th
                    class="px-6 py-3 whitespace-nowrap text-center"
                    colspan="1"
                  >
                    Notifications
                  </th>
                </tr>
              </thead>
              <tbody
                class="font-sans text-sm"
              >
                <tr
                  class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
                >
                  <td
                    class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
                  >
                    Alert 1
                  </td>
                </tr>
                <tr
                  class="border-b-2 bg-gray-100 border-gray-200 text-gray-700"
                >
                  <td
                    class="px-6 py-3 bg-white bg-opacity-70 whitespace-nowrap text-left"
                  >
                    Alert 2
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Notifications should not show an alert notification 1`] = `
<div>
  <button
    aria-expanded="false"
    class="p-1 rounded-full focus:outline-none focus:ring-2 focus:ring-offset-2 relative bg-gray-900 text-gray-400 focus:ring-offset-gray-900 hover:text-gray-100 focus:ring-gray-500"
    id="headlessui-popover-button-3"
    type="button"
  >
    <span
      class="sr-only"
    >
      Notifications
    </span>
    <svg
      aria-hidden="true"
      class="h-6 w-6"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15 17h5l-1.405-1.405A2.032 2.032 0 0118 14.158V11a6.002 6.002 0 00-4-5.659V5a2 2 0 10-4 0v.341C7.67 6.165 6 8.388 6 11v3.159c0 .538-.214 1.055-.595 1.436L4 17h5m6 0v1a3 3 0 11-6 0v-1m6 0H9"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
  </button>
</div>
`;

exports[`Notifications should show an alert notification 1`] = `
<div>
  <button
    aria-expanded="false"
    class="p-1 rounded-full focus:outline-none focus:ring-2 focus:ring-offset-2 relative bg-gray-900 text-gray-400 focus:ring-offset-gray-900 hover:text-gray-100 focus:ring-gray-500"
    id="headlessui-popover-button-1"
    type="button"
  >
    <span
      class="sr-only"
    >
      Notifications
    </span>
    <svg
      aria-hidden="true"
      class="h-6 w-6"
      fill="none"
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15 17h5l-1.405-1.405A2.032 2.032 0 0118 14.158V11a6.002 6.002 0 00-4-5.659V5a2 2 0 10-4 0v.341C7.67 6.165 6 8.388 6 11v3.159c0 .538-.214 1.055-.595 1.436L4 17h5m6 0v1a3 3 0 11-6 0v-1m6 0H9"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      />
    </svg>
    <span
      class="rounded-full text-sm font-medium bg-red-800 text-red-100 absolute top-0 right-0 p-0.5"
    >
      <span
        class="sr-only"
      >
        (
        New
        )
      </span>
      <svg
        aria-hidden="true"
        class="h-2.5 w-2.5"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
        />
      </svg>
    </span>
  </button>
</div>
`;
